### PR TITLE
feat(attachment-carousel): Image thumbnail with progress bar for the attachment carousel.

### PR DESCRIPTION
**Pull request by sanjeevdialpad**
# feat(attachment-carousel): Image thumbnail with progress bar for the attachment carousel.

## :hammer_and_wrench: Type Of Change

- [ ] Fix
- [x] Feature
- [ ] Refactoring
- [ ] Documentation

## :book: Description

These changes are specific to file types pertaining to images. It includes the progress bar when upload is in progress and also the close button which sends an event for the parent to handle removing the image.

This component uses DtImageViewer for the purpose of previewing the image.

### CODE REVIEW

File: recipes/conversation_view/attachment_carousel/attachment_carousel.vue [hunk 1]
```diff
@@ -191,8 +167,19 @@ export default {
   },
 
   methods: {
+    mediaComponent (type) {
+      switch (type) {
+        case 'image':
+          return 'dt-image-carousel';
+        default:
+          return 'dt-image-carousel';
+      }
```
Comment (braddialpad) [inline recipes/conversation_view/attachment_carousel/attachment_carousel.vue]: returns the exact same thing? whats the point?

File: recipes/conversation_view/attachment_carousel/attachment_carousel.vue [hunk 2]
```diff
@@ -191,8 +167,19 @@ export default {
...
+    },
+
     removeMediaItem (index) {
-      this.$emit('remove-image', index);
+      // make sure the carousel arrows is updated. 64 is the width of each media item
+      this.showRightArrow = this.$refs.carousel.scrollWidth > (this.$refs.carousel.clientWidth + 64);
```
Comment (braddialpad) [inline recipes/conversation_view/attachment_carousel/attachment_carousel.vue]: Remove magic number "64". Make it a constant called MEDIA_ITEM_WIDTH or something like that.

File: recipes/conversation_view/attachment_carousel/media_components/image_carousel.vue [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      class="dt-image-carousel--top-right"
+    >
+      <dt-progress-bar
+        v-if="mediaItem.isUploading && !showCloseButton"
+        class="dt-image-carousel--progress-bar d-ba d-bc-subtle d-baw2"
+        :progress="mediaItem.progress"
+      />
+      <dt-button
+        v-else
+        v-show="showCloseButton"
```
Comment (braddialpad) [inline recipes/conversation_view/attachment_carousel/media_components/image_carousel.vue]: bit confused with the v-else / v-show thing we got goin on here. Can we not just v-else? what are we working around here?

Comment (sanjeevdialpad) [reply recipes/conversation_view/attachment_carousel/media_components/image_carousel.vue]: on hover of the image, we have to show close button irrespective of if its uploading. 

The close button will be shown at all times but will be hidden if its not hovered. The uploading svg will only come up if that isUploading prop is set.

File: recipes/conversation_view/attachment_carousel/attachment_carousel.vue (hunk 1 above)
Comment (sanjeevdialpad) [reply recipes/conversation_view/attachment_carousel/attachment_carousel.vue]: yea i think we discussed to ignore and not show anything for type other than video/image. let me change that.

Comment (sanjeevdialpad) [reply recipes/conversation_view/attachment_carousel/attachment_carousel.vue]: i might have to preprocess the mediaList to remove any type other than image/video.

File: recipes/conversation_view/attachment_carousel/media_components/image_carousel.vue (hunk 3 above)
Comment (braddialpad) [reply recipes/conversation_view/attachment_carousel/media_components/image_carousel.vue]: so

```
<dt-progress-bar v-if="mediaItem.isUploading && !showCloseButton" />
<dt-button v-else-if="showCloseButton" />
```

File: recipes/conversation_view/attachment_carousel/media_components/image_carousel.vue [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        class="dt-image-carousel--close-button d-ba d-bc-white d-baw2"
```
Comment (braddialpad) [inline recipes/conversation_view/attachment_carousel/media_components/image_carousel.vue]: just add these utility classes to dt-image-carousel--close-button

File: recipes/conversation_view/attachment_carousel/media_components/image_carousel.vue [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  <div
+    role="presentation"
+    class="d-ps-relative"
+    @focusin="showCloseButton = true"
+    @focusout="showCloseButton = false"
+    @mouseenter="showCloseButton = true"
+    @mouseleave="showCloseButton = false"
+  >
+    <dt-image-viewer
+      image-button-class="d-h64 d-w64 d-bar4 d-ba d-baw6 d-bc-subtle"
```
Comment (braddialpad) [inline recipes/conversation_view/attachment_carousel/media_components/image_carousel.vue]: Create a new class for this

File: recipes/conversation_view/attachment_carousel/media_components/image_carousel.vue [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+      :aria-label="clickToOpenAriaLabel"
+    />
+
+    <!-- Loader / Close button -->
+    <div
+      class="dt-image-carousel--top-right"
+    >
+      <dt-progress-bar
+        v-if="mediaItem.isUploading && !showCloseButton"
+        class="dt-image-carousel--progress-bar d-ba d-bc-subtle d-baw2"
```
Comment (braddialpad) [inline recipes/conversation_view/attachment_carousel/media_components/image_carousel.vue]: add these util classes to `dt-image-carousel--progress-bar`

**Comment by sanjeevdialpad** [comment]
@braddialpad done the changes and had a few merge conflicts to solve. ready for final review.

**Review by braddialpad — APPROVED** [review]
Just tried it out, super cool how it animates! Thanks!

**Pull request merged**